### PR TITLE
Fix net test to exit cleanly across macOS/Linux

### PR DESCRIPTION
**Pull request by brianegge**
## Summary
Makes the `hobbes-test --tests Net` run exit cleanly on both macOS and Linux. The existing code used detached threads, which on macOS caused `libc++abi: terminating` at process exit (tests all pass, but the binary returns non-zero).

### Changes
- **`lib/hobbes/events/events.C`**
  - Implement `addTimer` on macOS (was `throw "nyi for OSX"`), required by the `eventLoopShutdownWithStopF` test.
  - In `stepEventLoop`, when an explicit stopFn is supplied with no timeout, poll every 500ms instead of blocking indefinitely — so a cleanup signal can wake the loop on both Linux (epoll_wait) and macOS (kevent).
- **`test/Net.C`**
  - Switch the test server from `thread.detach()` to a joined `NetTestCleanup` destructor that signals `serverStop=true` and joins at process exit.
  - Harden startup-condvar waits with a predicate (`serverReady` flag) so spurious wakeups can't race, and notify on the port-exhaustion path so the main thread can't hang.

### Follow-ups (not in this PR)
- Replace the 500ms poll with eventfd / `EVFILT_USER` wakeup so the default stopFn (`[]{ return false; }`) stays in an infinite block.
- Implement repeat-on-false / `EV_DELETE` / closure reclamation for macOS `addTimer`.

## Test plan
- [ ] macOS: `hobbes-test --tests Net` passes and exits 0 (previously emitted `libc++abi: terminating`)
- [ ] Linux: no process-exit hang (previously the cleanup would hang in `join()` because the event loop blocked forever in epoll_wait)

### CODE REVIEW

File: test/Net.C [hunk 1]
```diff
@@ -52,39 +59,59 @@ int testServerPort() {
...
 }
+
+// Clean up server threads at process exit
+static struct NetTestCleanup {
+  ~NetTestCleanup() {
+    serverStop = true;
+    serverWithHostStop = true;
+    if (serverThread.joinable()) {
+      serverThread.join();
+    }
```
[the remark this answers is not in the document]
Comment (brianegge) [reply test/Net.C]: Fixed in 10dffc8: replaced the namespace-scope `NetTestCleanup` with `std::atexit(joinServerThreadsAtExit)` registered inside `c()` after `cc x` is constructed. atexit handlers run in reverse registration order interleaved with static destructors, so the join now runs before `cc x`'s destructor.

File: test/Net.C [hunk 2]
```diff
@@ -16,32 +17,38 @@ static cc &c() {
...
   }
   serverPort = -1;
+  serverReady = true;
+  lk.unlock();
+  serverStartup.notify_one();
 }
 
 int testServerPort() {
   if (serverPort < 0) {
     std::unique_lock<std::mutex> lk(serverMtx);
```
[the remark this answers is not in the document]
Comment (brianegge) [reply test/Net.C:64]: Fixed in 10dffc8: `testServerPort` now joins any prior joinable `serverThread` before reassigning, and resets `serverReady = false` for the new attempt.

File: test/Net.C [hunk 3]
```diff
@@ -52,39 +59,59 @@ int testServerPort() {
...
   }
   serverPortWithHost = -1;
+  serverWithHostReady = true;
+  lk.unlock();
+  serverWithHostStartup.notify_one();
 }
 
 int testServerWithHostPort(const std::string &host = "") {
   if (serverPortWithHost < 0) {
     std::unique_lock<std::mutex> lk(serverMtxWithHost);
```
[the remark this answers is not in the document]
Comment (brianegge) [reply test/Net.C:110]: Fixed in 10dffc8: same treatment for `testServerWithHostPort` — join prior joinable thread and reset `serverWithHostReady` before reassigning.

File: lib/hobbes/events/events.C [hunk 4]
```diff
@@ -100,8 +100,12 @@ bool stepEventLoop(int timeoutMS, const std::function<bool()>& stopFn) {
       timeoutMS = std::max(1, millis);
     }
 
+    // When stopFn is provided with no explicit timeout, poll every 500ms
+    // so the stop condition gets checked instead of blocking indefinitely.
+    int effectiveTimeoutMS = timeoutMS < 0 ? 500 : timeoutMS;
+
```
[the remark this answers is not in the document]
Comment (brianegge) [reply lib/hobbes/events/events.C:106]: Acknowledged as a tracked follow-up (not in this PR's scope). The 500ms poll is a pragmatic fix for the process-exit hang; a proper wakeup primitive (eventfd on Linux, EVFILT_USER on macOS) will let `runEventLoop()` with the default stopFn stay in an infinite block and only wake on an explicit signal. Keeping this PR minimal — its purpose is to unblock clean process exit on macOS/Linux.

File: lib/hobbes/events/events.C [hunk 5]
```diff
@@ -239,19 +243,30 @@ void registerInterruptHandler(const std::function<void()>& fn) {
 
 bool stepEventLoop(int timeoutMS, const std::function<bool()>& stopFn) {
   while (!stopFn()) {
+    // When a stop function is provided and no explicit timeout is set,
+    // use a 500ms poll interval so the stop condition gets checked
+    // periodically instead of blocking indefinitely in kevent().
+    int effectiveTimeoutMS = timeoutMS;
+    if (timeoutMS < 0) {
+      effectiveTimeoutMS = 500;
+    }
```
[the remark this answers is not in the document]
Comment (brianegge) [reply lib/hobbes/events/events.C:252]: Acknowledged as a tracked follow-up (not in this PR's scope). The 500ms poll is a pragmatic fix for the process-exit hang; a proper wakeup primitive (eventfd on Linux, EVFILT_USER on macOS) will let `runEventLoop()` with the default stopFn stay in an infinite block and only wake on an explicit signal. Keeping this PR minimal — its purpose is to unblock clean process exit on macOS/Linux.

File: lib/hobbes/events/events.C [hunk 6]
```diff
@@ -268,8 +283,22 @@ void runEventLoop(const std::function<bool()>& stopFn) {
...
+    f();
+  });
+
+  int ident = nextTimerIdent++;
+  struct kevent ke;
+  EV_SET(&ke, ident, EVFILT_TIMER, EV_ADD, 0, millisecInterval, (void*)c);
+  if (kevent(kqfd, &ke, 1, 0, 0, 0) == -1) {
+    delete c;
+    throw std::runtime_error("Failed to add timer to kqueue: " + std::string(strerror(errno)));
+  }
```
[the remark this answers is not in the document]
Comment (brianegge) [reply lib/hobbes/events/events.C:301]: Acknowledged as a tracked follow-up. Prior to this PR `addTimer` on macOS just threw `"nyi for OSX"`, so the `eventLoopShutdownWithStopF` test couldn't run at all. This is the minimal implementation to unblock that test; proper lifecycle (repeat-on-false semantics, `EV_DELETE`, closure reclamation via a per-thread timer map) will come as a separate change.

**Review by kelliott55 — APPROVED** [review]
